File: lersosa-python/src/app/application/file/service/file_service_impl.py
#  Copyright (c) 2025 Leyramu Group. All rights reserved.
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
#  This project (Lersosa), including its source code, documentation, and any associated materials, is the intellectual property of Leyramu. No part of this software may be reproduced, distributed, or transmitted in any form or by any means, including photocopying, recording, or other electronic or mechanical methods, without the prior written permission of the copyright owner, Miraitowa_zcx, except in the case of brief quotations embodied in critical reviews and certain other noncommercial uses permitted by copyright law.
#
#  For inquiries related to licensing or usage outside the scope of this notice, please contact the copyright holder at 2038322151@qq.com.
#
#  The author disclaims all warranties, express or implied, including but not limited to the warranties of merchantability and fitness for a particular purpose. Under no circumstances shall the author be liable for any special, incidental, indirect, or consequential damages arising from the use of this software.
#
#  By using this project, users acknowledge and agree to abide by these terms and conditions.


from fastapi import UploadFile

from app.application.file.service.command import FileUploadCmdExe
from app.client.file.api.file_service_i import FileServiceI
from app.domain.entity import GrpcEntity


# 文件业务层
class FileServiceImpl(FileServiceI):
    """文件服务业务层
        提供与文件服务端交互的业务方法封装
    """

    # 初始化
    def __init__(self):
        self._fileUploadCmdExe: FileUploadCmdExe = FileUploadCmdExe()

    # 文件上传
    async def file_upload(self, file_request: UploadFile) -> GrpcEntity:
        """执行文件上传命令执行器
        Args:
            file_request: UploadFile 包含待上传文件对象，包含文件名、内容类型和文件内容
        Returns:
            GrpcEntity: 包含gRPC服务返回的响应实体
        """
        return await self._fileUploadCmdExe.execute(file_request)

File: lersosa-java/lersosa-service/lersosa-service-grpc/lersosa-service-grpc-app/src/main/java/leyramu/framework/lersosa/grpc/file/service/FileServiceImpl.java
/*
 * Copyright (c) 2023-2025 Leyramu Group. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This project (Lersosa), including its source code, documentation, and any associated materials, is the intellectual property of Leyramu. No part of this software may be reproduced, distributed, or transmitted in any form or by any means, including photocopying, recording, or other electronic or mechanical methods, without the prior written permission of the copyright owner, Miraitowa_zcx, except in the case of brief quotations embodied in critical reviews and certain other noncommercial uses permitted by copyright law.
 *
 * For inquiries related to licensing or usage outside the scope of this notice, please contact the copyright holder at 2038322151@qq.com.
 *
 * The author disclaims all warranties, express or implied, including but not limited to the warranties of merchantability and fitness for a particular purpose. Under no circumstances shall the author be liable for any special, incidental, indirect, or consequential damages arising from the use of this software.
 *
 * By using this project, users acknowledge and agree to abide by these terms and conditions.
 */

package leyramu.framework.lersosa.grpc.file.service;

import io.grpc.stub.StreamObserver;
import leyramu.framework.lersosa.common.core.utils.SpringUtils;
import leyramu.framework.lersosa.common.grpc.annotation.GrpcLog;
import leyramu.framework.lersosa.common.grpc.lib.file.FileGrpc;
import leyramu.framework.lersosa.common.grpc.lib.file.FileOuterClass;
import leyramu.framework.lersosa.common.purge.annotation.NgxCacheCls;
import leyramu.framework.lersosa.common.tenant.helper.TenantHelper;
import leyramu.framework.lersosa.grpc.api.FileService;
import leyramu.framework.lersosa.grpc.file.command.FileSaveCmdExe;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.devh.boot.grpc.server.service.GrpcService;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.web.multipart.MultipartFile;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.List;

/**
 * 文件服务业务层 接口.
 *
 * @author <a href="mailto:2038322151@qq.com">Miraitowa_zcx</a>
 * @version 2.0.0
 * @since 2025/3/7
 */
@Slf4j
@GrpcService
@RequiredArgsConstructor
public class FileServiceImpl extends FileGrpc.FileImplBase implements FileService {

    /**
     * 文件保存命令执行器.
     */
    private final FileSaveCmdExe fileSaveCmdExe;

    /**
     * Grpc 上传文件.
     *
     * @param responseObserver 响应对象
     * @return 文件对象
     */
    @GrpcLog
    @Override
    public StreamObserver<FileOuterClass.FileRequest> uploadFile(StreamObserver<FileOuterClass.FileReply> responseObserver) {

        // TODO 动态设置租户ID
        TenantHelper.setDynamic("000000");
        FileServiceImpl proxy = SpringUtils.getAopProxy(this);

        return new StreamObserver<>() {

            /**
             * 缓存数据.
             */
            private final ByteArrayOutputStream buffer = new ByteArrayOutputStream();

            /**
             * 文件名.
             */
            private String fileName;

            /**
             * 文件类型.
             */
            private String mimeType;

            /**
             * 接收数据.
             *
             * @param request 文件请求
             */
            @Override
            public void onNext(FileOuterClass.FileRequest request) {
                FileOuterClass.FileChunk chunk = request.getFileChunk();
                try {
                    if (fileName == null) {
                        fileName = chunk.getFileName();
                        mimeType = chunk.getMimeType();
                    }
                    chunk.getData().writeTo(buffer);
                } catch (IOException e) {
                    responseObserver.onError(e);
                }
            }

            /**
             * 完成处理.
             */
            @Override
            public void onCompleted() {
                try {
                    MultipartFile multipartFile = new MockMultipartFile(
                        fileName,
                        fileName,
                        mimeType,
                        buffer.toByteArray()
                    );

                    List<FileOuterClass.FileDataReply> fileDataReply = fileSaveCmdExe.save(multipartFile).stream()
                        .map(
                            remoteFile -> FileOuterClass.FileDataReply.newBuilder()
                                .setName(remoteFile.getName())
                                .setUrl(remoteFile.getUrl())
                                .setOriginalName(remoteFile.getOriginalName())
                                .setFileSuffix(remoteFile.getFileSuffix())
                                .build()
                        )
                        .toList();

                    responseObserver.onNext(FileOuterClass.FileReply.newBuilder()
                        .setCode(200)
                        .setMessage("共计" + fileDataReply.size() + "文件上传成功")
                        .addAllData(fileDataReply)
                        .build());
                    proxy.cleanMarkListCache();
                    responseObserver.onCompleted();
                } catch (Exception e) {
                    responseObserver.onNext(
                        FileOuterClass.FileReply.newBuilder()
                            .setCode(500)
                            .setMessage("文件上传失败，存在重复项！")
                            .build()
                    );
                    responseObserver.onCompleted();
                    log.error("文件上传失败", e);
                }
            }

            /**
             * 错误处理.
             * @param t 异常
             */
            @Override
            public void onError(Throwable t) {
                log.error("文件上传失败", t);
            }
        };
    }

    /**
     * 清理缓存.
     */
    @CacheEvict(
        cacheNames = {"pulsar:mark:list", "pulsar:mark:chartE"},
        allEntries = true,
        beforeInvocation = true
    )
    @NgxCacheCls
    public void cleanMarkListCache() {
        TenantHelper.clearDynamic();
        log.info("Redis 缓存清理成功");
    }
}
